Unix/http/httpclientauth.c

```diff
@@ -1359,33 +1384,105 @@ HttpClient_EncryptData(_In_ HttpClient_SR_SocketData * handler, _Out_ Page **pHe
 
     gss_buffer_desc input_buffer = { original_content_len, poriginal_data };
     gss_buffer_desc output_buffer = { 0 };
     OM_uint32 min_stat, maj_stat;
     int out_flags;
+    OM_uint32 signature_length = 16;  // NTLM_SIGNATURE_LENGTH
 
-    maj_stat = (*_g_gssClientState.Gss_Wrap)(&min_stat, handler->authContext, (handler->negoFlags & (GSS_C_INTEG_FLAG | GSS_C_CONF_FLAG)),
-                        GSS_C_QOP_DEFAULT, &input_buffer, &out_flags, &output_buffer);
+    char *alloced_data = NULL;  // We will have a leak if we don't free the memory we alloc for the output buffer. 
+
+#if GSS_USE_IOV
+
+    if (_g_gssClientState.Gss_Wrap_Iov && handler->selectedMech == AUTH_MECH_KERBEROS)
+    {
+        gss_iov_buffer_desc iov[4] = {{0}};
+
+        // We can get here either by requesting Kerberos directly, or negotiating it via Spnego
+        // Either way, MIT gss_wrap and SSPI EncryptMessage don't get along.
+        //
+        iov[0].type = GSS_IOV_BUFFER_TYPE_HEADER     | GSS_IOV_BUFFER_FLAG_ALLOCATE;
+        iov[1].type = GSS_IOV_BUFFER_TYPE_DATA;
+        iov[1].buffer.length = original_content_len;
+        iov[1].buffer.value  = poriginal_data;
+
+        maj_stat = (*_g_gssClientState.Gss_Wrap_Iov)(&min_stat, handler->authContext, (handler->negoFlags & GSS_C_CONF_FLAG),
+                            GSS_C_QOP_DEFAULT, &out_flags, iov, 2);
+        if (maj_stat != GSS_S_COMPLETE)
+        {
+            _ReportError(handler, "gss_wrap failed", maj_stat, min_stat);
+            (*_g_gssClientState.Gss_Release_Iov_Buffer)(&min_stat, iov, 2);
+            return MI_FALSE;
+        }
 
+    // We need to get the proper length from the token, but for now 16 bytes is correct for all 
+    // protocols
+
+        signature_length = iov[0].buffer.length;
+
+        output_buffer.length = (iov[0].buffer.length+
+                                iov[1].buffer.length
+                                /* +iov[2].buffer.length+iov[3].buffer.length*/ );
+
+        alloced_data = PAL_Malloc(output_buffer.length); // We retain this so we can tell we alloced it, not gss
+        output_buffer.value = alloced_data;
+    
+        char *outbufp = output_buffer.value;
+        if (iov[0].buffer.length)
+        {
+            memcpy(outbufp, iov[0].buffer.value, iov[0].buffer.length);
+            outbufp += iov[0].buffer.length;
+        }
+    
+        if (iov[1].buffer.length)
+        {
+            memcpy(outbufp, iov[1].buffer.value, iov[1].buffer.length);
+            outbufp += iov[1].buffer.length;
+        }
+
+        (*_g_gssClientState.Gss_Release_Iov_Buffer)(&min_stat, iov, 2);
+
+    }
+    else 
+    {
+        if (handler->selectedMech == AUTH_MECH_NTLMSSP)
+        {
+            //  We are here because gss_wrap_iov is not implemented in gss-ntlmssp. 
+            //  If we aren't doing NTLM I'm not sure how we would figure out the signature length
+            
+             signature_length = 16;  // NTLM_SIGNATURE_LENGTH
```

Comment:
Kerberos has a different siglen, but it is not constant.  NTLM is always 16, so we preinit to 16 and change it if we are doing kerb.